Sources/Private/Experimental/Layers/CALayer+setupLayerHierarchy.swift

```diff
@@ -0,0 +1,71 @@
+// Created by Cal Stephens on 1/11/22.
+// Copyright © 2022 Airbnb Inc. All rights reserved.
+
+import QuartzCore
+
+extension CALayer {
+  /// Sets up an `AnimationLayer` / `CALayer` hierarchy in this layer,
+  /// using the given list of layers.
+  func setupLayerHierarchy(
+    for layers: [LayerModel],
+    context: LayerContext)
+  {
+    // An `Animation`'s `LayerModel`s are listed from front to back,
+    // but `CALayer.sublayers` are listed from back to front.
+    // We reverse the layer ordering to match what Core Animation expects.
+    // The final view hierarchy must display the layers in this exact order.
+    let layersInZAxisOrder = layers.reversed()
+
+    let layersByIndex = Dictionary(grouping: layersInZAxisOrder, by: \.index)
+      .compactMapValues(\.first)
+
+    /// Layers specify a `parent` layer. Child layers inherit the `transform` of their parent.
+    ///  - We can't add the child as a sublayer of the parent `CALayer`, since that would
+    ///    break the ordering specified in `layersInZAxisOrder`.
+    ///  - Instead, we create an invisible `TransformLayer` to handle the parent
+    ///    transform animations, and add the child layer to that `TransformLayer`.
+    func setupParentTransformLayers(
```

Comment:
The previous setup, where we would add children to their parent using `addSublayer`, was incorrect.

To render animations correctly, we need to exactly respect `layersInZAxisOrder`. It's possible to create orderings that are incompatible with an actual sublayer hierarchy. For example, if you have `[A, B, C]` where `C` is a child of `A`, then adding `C` as a sublayer of `A` would incorrectly change the layer ordering to `[A, C, B]`.

Instead, we now create invisible `TransformLayer`s to display / animate the `transform` of the parent. This makes sure the child layer correctly inherits the parent transform without affecting layer ordering.

This fixes a bunch of issues with the sample animations. Some examples:

| Before | After |
| ----- | ---- |
| ![Simulator Screen Shot - iPhone 8 - 2022-01-11 at 12 45 44](https://user-images.githubusercontent.com/1811727/149021892-98c090c0-af7f-4b17-a750-e5bb31d9c9ff.png) | ![Simulator Screen Shot - iPhone 8 - 2022-01-11 at 12 46 49](https://user-images.githubusercontent.com/1811727/149021900-dc8b0252-d6ac-462e-99bc-88b7a28747df.png) |
| ![Simulator Screen Shot - iPhone 8 - 2022-01-11 at 12 46 16](https://user-images.githubusercontent.com/1811727/149021922-3bf07e30-e978-4022-9218-96db4b259331.png) | ![Simulator Screen Shot - iPhone 8 - 2022-01-11 at 12 47 24](https://user-images.githubusercontent.com/1811727/149021928-568ef2e3-c95d-4cae-8bed-6c1f28671b3b.png) |

Reply:
Great solve!